### PR TITLE
CRM-19447 activity details search

### PR DESCRIPTION
**Pull request by twomice**
This applies the requested changes from CRM-19447, on both the Activity Search and the Advanced Search.

---

 * [CRM-19447: Improve Activity Subject\/Details text search to be similiar to Notes Subject\/Body](https://issues.civicrm.org/jira/browse/CRM-19447)

### CODE REVIEW

**Comment by monishdeb** [comment]
@twomice you can also add unit test on search form's ```formValues```. Lemme cite an example https://github.com/civicrm/civicrm-core/pull/9506/files#diff-ebb82d7b9ab678b644e21a15f37a8e86R162

File: CRM/Activity/BAO/Query.php [hunk 1]
```diff
@@ -190,15 +190,14 @@ public static function whereClauseSingle(&$values, &$query) {
       case 'activity_type_id':
       case 'activity_status_id':
       case 'activity_engagement_level':
-      case 'activity_subject':
```
Comment (monishdeb) [inline CRM/Activity/BAO/Query.php]: Although there is a api test to assert ```subject``` but will it cause any unintended regression ?

File: CRM/Activity/Form/Search.php [hunk 2]
```diff
@@ -195,7 +195,6 @@ public function postProcess() {
       $specialParams = array(
         'activity_type_id',
         'status_id',
-        'activity_subject',
```
Comment (monishdeb) [inline CRM/Activity/Form/Search.php]: you can replace it with activity_text``` as ```CRM_Contact_BAO_Query::processSpecialFormValue(...)``` can format LIKAble attributes here https://github.com/twomice/civicrm-core/blob/96e30da363f70bb7919276a87567892cfad0d60b/CRM/Contact/BAO/Query.php#L6073 and eventually it optimise ```whereClauseSingleActivityText(...)``` to handle all kind of operator not just hardcoded LIKE

**Comment by twomice** [comment]
Thanks, @monishdeb. I will look at these.

**Comment by twomice** [comment]
@monishdeb Regarding "you can also add unit test on search form's formValues": from the example, I'm not sure what I'm really testing, and accordingly where to locate the test (e.g., tests/phpunit/CRM/Contact/BAO/...what?) Can you offer any insight?

**Comment by monishdeb** [comment]
@twomice here's few example how you can add unit-test on Search Form's ```form-values``` and even Qill string too. 
1. using Smart and/or Regular group filter [testContactCombineGroupSearch](https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/CRM/Contact/BAO/GroupContactTest.php#L162) in Advance Search
2. using Group Type filter [testbyGroupType](https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/CRM/Contact/BAO/GroupContactTest.php#L238) in Advance Search
3. using 'Is Batch' filter [testBatchFilter](https://github.com/civicrm/civicrm-core/pull/9492/files#diff-c4448e6d970120e7c10d0a05bf6008b5R53) in Find Contribution .

**Comment by twomice** [comment]
Sorry to ask @monishdeb, but Jenkins testing is failing to merge the PR now:

  error: patch failed: CRM/Activity/BAO/Query.php:192                                
  error: CRM/Activity/BAO/Query.php: patch does not apply                            
  error: patch failed: CRM/Activity/Form/Search.php:195                              
  error: CRM/Activity/Form/Search.php: patch does not apply
(https://test.civicrm.org/job/CiviCRM-Core-PR/13156/console)

Any suggestions what I can do to fix this?

**Comment by monishdeb** [comment]
@twomice I have updated the PR and test build passes :) 

U might have missed --rebase, reason why there was a false merge commit. 
Always do ```git pull --rebase upstream master``` to update your PR and in order to squash multiple commits into one do ```git rebase -i``` .. it will open a interactive console in vi edit.

**Comment by twomice** [comment]
Thanks @monishdeb! Let me know if the PR needs more work.

**Comment by monishdeb** [comment]
@twomice will do the QA. However some might want to have separate search filters for Activity ```details``` and ```subject``` respectively for sake of simplicity. But the current feature gives the ability to use the same filter for activity detail and/or subject.

**Comment by twomice** [comment]
@monishdeb The intent of the functionality as describedin the ticket CRM-19447 is to emulate the behavior seen for Notes, as in Advanced Search > Notes > Note Text, with options "Body Only / Subject Only / Both".  So this PR aims to do what's called for in the ticket. Hopefully that's acceptable.

**Comment by monishdeb** [comment]
Oh ok, very well proceeding with QA. There are few scope for optimization and will update this PR with those changes.

**Comment by monishdeb** [comment]
@twomice did some additional changes https://github.com/civicrm/civicrm-core/pull/9536/commits/1cd7479eddd0eb97cf009e182fe571065df08aa9 please have a look. Will merge then after test builds passes

**Comment by monishdeb** [comment]
image

Top of the changes

**Comment by twomice** [comment]
Those are great changes, @monishdeb Thanks for that, and for the merge!

**Pull request merged**